artiq/gateware/rtio/phy/edge_counter.py

```diff
@@ -0,0 +1,83 @@
+from migen import *
+from artiq.gateware.rtio import rtlink
+
+
+class SimpleEdgeCounter(Module):
+    """Counts rising/falling edges of an input signal.
+
+    Control (sensitivity/zeroing) is done via a single RTIO output channel,
+    which is is also used to request an input event to be emitted with the
+    current counter value.
+
+    :param input_state: The (scalar) input signal to detect edges of.
+    :param counter_width: The width of the counter register, in bits. Defaults
+        to 31 to match integers being signed in ARTIQ Python.
+    """
+
+    def __init__(self, input_state, counter_width=31):
+        assert counter_width >= 2
+
+        # RTIO interface:
+        #  - output 0: 4 bits, <count_rising><count_falling><send_event><zero_counter>
+        #  - input 0: 32 bits, accumulated edge count
+        self.rtlink = rtlink.Interface(
+            rtlink.OInterface(5, enable_replace=False),
+            rtlink.IInterface(counter_width))
+
+        #: Registered copy of the input state, in the rio clock domain.
+        self.input_state = Signal()
+
+        # # #
+
+        current_count = Signal(counter_width)
+
+        count_rising = Signal()
+        count_falling = Signal()
+        send_event_stb = Signal()
+        zero_counter_stb = Signal()
+
+        # Read configuration from RTIO output events.
+        self.sync.rio_phy += [
+            If(self.rtlink.o.stb, count_rising.eq(self.rtlink.o.data[0]),
+               count_falling.eq(self.rtlink.o.data[1]),
+               send_event_stb.eq(self.rtlink.o.data[2]),
+               zero_counter_stb.eq(self.rtlink.o.data[3])).Else(
+                   send_event_stb.eq(0),
+                   zero_counter_stb.eq(0),
+               )
+        ]
+
+        # Generate RTIO input event with current count if requested.
+        event_data = Signal.like(current_count)
+        self.comb += [
+            self.rtlink.i.stb.eq(send_event_stb),
+            self.rtlink.i.data.eq(event_data)
+        ]
+
+        # Register input state, keeping previous one for edge detection.
+        input_state_d = Signal()
+        self.sync.rio += [
+            self.input_state.eq(input_state),
+            input_state_d.eq(self.input_state)
+        ]
+
+        # Count input edges, saturating at the maximum.
+        new_count_raw = Signal(counter_width + 1)
+        self.comb += new_count_raw.eq(
+            Cat(current_count, C(0, 1)) +
+            (count_rising & (self.input_state & ~input_state_d)) +
+            (count_falling & (~self.input_state & input_state_d)))
+        new_count = Signal.like(current_count)
+        self.comb += new_count.eq(
+            Mux(new_count_raw[counter_width], 2**counter_width - 1,
```

Comment:
Is saturation really important here? One could argue that overflow is well defined for a counter in terms of aliasing.
Also the saturation logic as written seems convoluted.

Reply:
There should at least be something to tell you that overflow has occured.

Reply:
Ok. Could just use the msb for that (set on overflow, clear on counter clear).